[WIP] Transformer Chainer

Fhrozen opened the pull request:
The initial commit to support transformer in chainer. This is a extension of #465 which will be closed.

Some detail:
- It is most compatible with pytorch code, but some functions are required to be modified.
- Added test but not for chainer (currently)
- Because of the implementation is no longer recurrent, the use of `loss.unchain_backward()` seems unnecessary in the training of the transformer. It also seems to affect the training time (Need more test). If this is removed during training the transformer, then the visualization of the full network becomes possible. Let me know if this is required.
- The model only trains in att mode, and decodes with greedy search.  Additional details are WIP.

I am testing it to submit the initial results. The log says that the training will be finishing in a 33 hrs (quite larger than any additional test I did).

sw005320 commented:
There seem to have some bash script issues in https://travis-ci.org/espnet/espnet/jobs/510541041#L1341-L1350
Please fix them.

@gtache, the bash script check (https://github.com/espnet/espnet/blob/master/.travis.yml#L77) does not seem to exit with the appropriate error code. Could you check it?

sw005320 commented:
We first merge this, and move to the pytorch backend.

ShigekiKarita commented on egs/wsj/asr2/run.sh:
```diff
@@ -0,0 +1,370 @@
+        asr_train.py \
+        --accum-grad ${accum_grad} \
+        --ngpu ${ngpu} \
+        --backend ${backend} \
```
I think this `--backend ${backend}` and `--model-module "espnet.nets.${backend}_backend.asr_transformer" ` are redundant. how about identifing the backend in `asr_train.py` like
```python
module = load_module(args.model_module)
args.backend = "chainer" if isinstance(module.E2E, chainer.Chain) else "pytorch"
```

ShigekiKarita commented on egs/wsj/asr2/run.sh:
```diff
@@ -0,0 +1,370 @@
+        --epochs ${epochs} \
+        --sortagrad ${sortagrad} \
+        --lsm-weight ${lsm_weight} \
+        --model-module "espnet.nets.${backend}_backend.asr_transformer" \
```
It is not user-friendly to assume implicitly that `E2E` class is defined in this module. How about specifying explicitly `espnet.nets.${backend}_backend.asr_transformer.E2E`  instead?

ShigekiKarita commented on egs/wsj/asr2/run.sh:
```diff
@@ -0,0 +1,370 @@
+
+# network architecture
+ninit=pytorch
+ntype=transformer
```
I think $ntype is only used in $expdir and it does not change `asr_train.py` behavior. How about new cmd arg `mode_lmodule="espnet.nets.${backend}_backend.asr_transformer.E2E"` instead to indicate the exact model?

ShigekiKarita commented on espnet/asr/chainer_backend/asr.py:
```diff
@@ -104,34 +108,42 @@ def update_core(self):
-        else:
+        elif self.count % self.accum_grad == 0:
             optimizer.update()
+            self.clean = True
```
why don't you clean up gradients immediately? I mean how about `optimizer.target.cleargrads()` here.

ShigekiKarita commented on espnet/asr/chainer_backend/asr.py:
```diff
@@ -154,8 +166,9 @@ def update_core(self):
-            else:
+            elif self.count % self.accum_grad == 0:
                 optimizer.update()
+                self.clean = True
```
ditto

ShigekiKarita commented:
LGTM except for several comments above. For `--model_module` argument, I think following existing notation in the `transformation` module is the best.

Fhrozen commented:
@ShigekiKarita san, thank you for your hard work and the comments.
I will submitting the PR with the modifications by tomorrow.

sw005320 commented on egs/wsj/asr2/run.sh:
```diff
@@ -0,0 +1,370 @@
+        --dunits ${dunits} \
+        --aheads ${aheads} \
+        --adim ${adim} \
+        --mtlalpha ${mtlalpha} \
```
We can keep this, but please add some WARNING in the python program that this is not implemented yet.

Fhrozen replied:
The code internally raise an NotImplemented Flag:

https://github.com/Fhrozen/espnet/blob/0b77d8d39b850b4a6373ff0feae6c324a76e28f4/espnet/nets/chainer_backend/asr_transformer/e2e.py#L106-L107

Should I change it for a warning?

sw005320 replied:
OK, thanks

sw005320 commented on egs/wsj/asr2/run.sh:
```diff
@@ -0,0 +1,370 @@
+            --recog-json ${feat_recog_dir}/split${nj}utt/data.JOB.json \
+            --result-label ${expdir}/${decode_dir}/data.JOB.json \
+            --model ${expdir}/results/${recog_model}  \
+            --beam-size ${beam_size} \
```
We can keep this, but please add some WARNING in the python program that this is not implemented yet.

Fhrozen replied:
Same here:

https://github.com/Fhrozen/espnet/blob/0b77d8d39b850b4a6373ff0feae6c324a76e28f4/espnet/nets/chainer_backend/asr_transformer/e2e.py#L306-L307

Using raiseError

sw005320 commented on egs/wsj/asr2/run.sh:
```diff
@@ -0,0 +1,370 @@
+            --penalty ${penalty} \
+            --maxlenratio ${maxlenratio} \
+            --minlenratio ${minlenratio} \
+            --ctc-weight ${ctc_weight} \
```
We can keep this, but please add some WARNING in the python program that this is not implemented yet.

sw005320 commented on egs/wsj/asr2/run.sh:
```diff
@@ -0,0 +1,370 @@
+            --maxlenratio ${maxlenratio} \
+            --minlenratio ${minlenratio} \
+            --ctc-weight ${ctc_weight} \
+            --lm-weight ${lm_weight} \
```
We can keep this, but please add some WARNING in the python program that this is not implemented yet.

sw005320 commented on espnet/nets/chainer_backend/asr_transformer/attention.py:
```diff
@@ -0,0 +1,60 @@
+# encoding: utf-8
```
This function must be at `espnet/espnet/nets/chainer_backend/`
How about `espnet/espnet/nets/chainer_backend/attention_transformer.py`?

sw005320 commented on espnet/nets/chainer_backend/asr_transformer/__init__.py:
```diff
@@ -0,0 +1,4 @@
+from espnet.nets.chainer_backend.asr_transformer.e2e import E2E  # NOQA
```
I'm not a big fan of making `asr_transformer` directory. I think `asr` directory should be OK.

sw005320 commented on espnet/nets/chainer_backend/asr_transformer/decoders.py:
```diff
@@ -0,0 +1,79 @@
+# encoding: utf-8
```
This function must be at `espnet/espnet/nets/chainer_backend/`
How about `espnet/espnet/nets/chainer_backend/decoders_transformer.py`?

sw005320 commented on espnet/nets/chainer_backend/asr_transformer/e2e.py:
```diff
@@ -0,0 +1,328 @@
+# encoding: utf-8
```
This function must be at `espnet/espnet/nets/chainer_backend/`.
How about `espnet/espnet/nets/chainer_backend/e2e_asr_transformer.py`?

sw005320 commented on espnet/nets/chainer_backend/asr_transformer/encoders.py:
```diff
@@ -0,0 +1,118 @@
+# encoding: utf-8
```
How about `espnet/espnet/nets/chainer_backend/encoders_transformer.py`?

sw005320 commented on espnet/nets/chainer_backend/asr_transformer/net_utils.py:
```diff
@@ -0,0 +1,171 @@
+# encoding: utf-8
```
Add this one at `espnet/espnet/nets/chainer_backend/nets_utils_transformer.py`

Fhrozen commented:
Currently testing the new code:
Some points to resolve (or modify):
- [ ] Add results
- [ ] Add transformer test for chainer
- [ ] Dynamic loading. I am honestly not sure about this part because it will require to modify all recipes, Currently, the modifications just affect the transformer recipe. Maybe this could target in different PR.

ShigekiKarita commented:
> Dynamic loading
OK. Maybe I will handle this in the pytorch PR later.

Fhrozen commented:
I just finished an additional commit:
- Trained and decoded for the large model and added the results.
- Added a test for chainer transfomer (train linear and decode). I am not sure about the other tests.
- resolved the cleargrad part. It seems that when training with single GPU and the grads are not cleared at the beginning, then the model does not update (grad_norm is nan). So I just added a cleargrad if count == 1, For multigpu the training seems to have no problem.

Let me know if there is any additional point to resolve or fix.

ShigekiKarita commented:
LGTM :checkered_flag:

sw005320 commented:
Thanks @Fhrozen!!!
@ShigekiKarita, please follow up it!
